Source/missiles.cpp

```diff
@@ -4015,11 +4005,21 @@ void MI_Lightctrl(int i)
 				    dam,
 				    missile[i]._mispllvl);
 			}
-			missile[i]._miVar1 = missile[i].position.tile.x;
-			missile[i]._miVar2 = missile[i].position.tile.y;
+		} else {
+			AddMissile(
+			    missile[i].position.tile,
+			    missile[i].position.start,
+			    i,
+			    MIS_LIGHTNING,
+			    missile[i]._micaster,
+			    missile[i]._misource,
+			    dam,
+			    missile[i]._mispllvl);
 		}
+		missile[i]._miVar1 = missile[i].position.tile.x;
+		missile[i]._miVar2 = missile[i].position.tile.y;
 	}
-	if (missile[i]._mirange == 0 || mx <= 0 || my <= 0 || mx >= MAXDUNX || my > MAXDUNY) {
+	if (missile[i]._mirange == 0 || mx == 0 || my == 0) {

```

Comment:
If the checks I've deleted aren't redundant, then accessing dPiece[mx][my] is undefined behavior.

Reply:
In other words, is there a bug here?

Reply:
technically yes but actually no (xD) as even if accessing dPiece[mx][my] is undefined behavior, the condition after it checks for the bounds so nothing scary happens but stilll ... maybe move the bounds check in front of the dpiece?

Reply:
There's "int pn = dPiece[mx][my]", and then some pn-related stuff right at the top of the function.

Reply:
It looks like the x/y == 0 checks are to ensure that the missile can't move out of bounds as a workaround for the out of order/missing bounds check before indexing into dPiece you've pointed out. ProcessMissiles() checks if the missile is still in bounds (using a buggy check of `>= MAXDUNX/y - 1` where either `> MAXDUNX - 1` or `>= MAXDUNX` would've been appropriate) before calling this function, the missile position should be in the range (0,0) to (MAXDUNX-2, MAXDUNY-2) as a result. The call to GetMissilePos causes the position to be recalculated based on the start location and the distance travelled so far, but I *think* this results in the same position value.

Reply:
The point I was making is that dPiece is MAXDUNX x MAXDUNY in size, and mx and my are used to access one of its elements. For it not to be UB, mx and my must be in [0, MAXDUNX/Y -1].

mx and my are never mutated throughout the function, and the bounds checking is done at the very end, i.e. too late to prevent any potential bugs when accessing dPiece if the values are out of range.

I don't know if we can assume that the values supplied ([mx, my] == initial value of missile[i].position.tile) are always within bounds. Hence, the late bounds check tipped me off that there might be some bug lurking here.

Reply:
Apologies, I didn't explain that particularly well, I was trying to say that the check at the end of that missile procedure guarantees that the missile does not move out of bounds next time it is called. However I couldn't be certain about the initial conditions. Looking into it further the flow for this type of missile is:

1. AddMissile() initialises the missile position and start location with whatever values were passed in. For MIS_LIGHTCTRL missiles this is only called from MI_Chain with either the current position of a valid missile (or the casting entity) or a position that has been checked to fall within (0, MAXDUNX), (0, MAXDUNY) - any valid dungeon tile except row 0 or col 0
2. AddLightctrl() sets the initial velocity (where x is in the range [-32, 32] and y is in the range [-16, 16])
3. ProcessMissiles() checks the missile position is "valid", meaning x is within [0, MAXDUNX-2]and y is within [0, MAXDUNY-2]
    * Missiles spawned in or which travelled to the last row/column of the dungeon are deleted
4. MI_Lightctrl() updates how far the missile has travelled using travelled += velocity and GetMissilePos()
    * due to the way the velocity is set initially this moves the missile a number of subpixels, this appears to result in movement 1 tile in the x direction and 0.5 tiles in the y direction relative to the last update
    * The missile cannot move past MAXDUNX/Y due to ProcessMissiles killing missiles in the last row/col. The asserts before indexing into dPiece should never be triggered
5. MI_Lightctrl() performs the missile logic, then flags the missile for deletion if it moved into row/col 0
    * This prevents missiles which have velocity {-32, 0} from resulting in a read out of dpiece[-1][y] in the next update

It looks like there is a potential bug if a player (or monster?) casts chain lightning to the -x direction while standing in col 0 (or in the -y direction while standing in row 0). I haven't checked the map generation code to see if it's possible for a dungeon layer to generate with an empty border tile, and I haven't checked any of the other missile types for similar bugs.

Reply:
> I haven't checked the map generation code to see if it's possible for a dungeon layer to generate with an empty border tile
 
I have seen screenshots of a few seeds where dungeon generation lead outside of the bounds (I think into negative Y), and it causes a hall of mirrors effect on the area that is open. I can't remember if there are bad effects to missiles or not, I'm trying to locate information on the seeds as it's been a while since I've seen it. If I can find it, I'll either update this or add a new comment with it.